app/com/arpnetworking/kairos/service/KairosDbServiceImpl.java

```diff
@@ -0,0 +1,256 @@
+                        _metricsList.set(list);
+                        return list;
+                    });
+
+            final List<String> metricsList = _metricsList.get();
+            if (metricsList != null) {
+                response = CompletableFuture.completedFuture(metricsList);
+            } else {
+                response = queryResponse;
+            }
+        }
+
+        return response;
+    }
+
+    private static MetricsQuery useAvailableRollups(final List<String> metricNames, final MetricsQuery originalQuery) {
+        final MetricsQuery.Builder newQueryBuilder = new MetricsQuery.Builder()
+                .setStartTime(originalQuery.getStartTime());
+
+        originalQuery.getEndTime().ifPresent(newQueryBuilder::setEndTime);
+
+
+        newQueryBuilder.setMetrics(originalQuery.getMetrics().stream().map(metric -> {
+            // Check to see if there are any rollups for this metrics
+            final String metricName = metric.getName();
+            final ImmutableList<String> filteredMetrics = filterMetricNames(metricNames, Optional.of(metricName), false);
+            final List<String> rollupMetrics = filteredMetrics
+                    .stream()
+                    .filter(IS_ROLLUP)
+                    .filter(s -> s.length() == metricName.length() + 3)
+                    .collect(Collectors.toList());
+
+            if (rollupMetrics.isEmpty()) {
+                // No rollups so execute what we received
+                return metric;
+            } else {
+                // There are rollups, now determine the appropriate one based on the max sampling period in the
+                // aggregators
+                final Optional<SamplingUnit> maxUnit = metric.getAggregators().stream()
+                        .filter(agg -> agg.getAlignSampling().orElse(Boolean.FALSE)) // Filter out non-sampling aligned
+                        .map(Aggregator::getSampling)
+                        .map(sampling -> sampling.map(Sampling::getUnit).orElse(SamplingUnit.MILLISECONDS))
+                        .max(SamplingUnit::compareTo);
+
+                // No aggregators are sampling aligned so skip as rollups are always aligned
+                if (maxUnit.isPresent()) {
+
+                    final TreeMap<SamplingUnit, String> orderedRollups = new TreeMap<>();
+                    rollupMetrics.forEach(name -> {
+                        final Optional<SamplingUnit> rollupUnit = rollupSuffixToSamplingUnit(name.substring(metricName.length() + 1));
+                        rollupUnit.ifPresent(samplingUnit -> orderedRollups.put(samplingUnit, name));
+                    });
+
+                    final Map.Entry<SamplingUnit, String> floorEntry = orderedRollups.floorEntry(maxUnit.get());
+                    final String rollupName = floorEntry != null ? floorEntry.getValue() : metricName;
+                    final Metric.Builder metricBuilder = new Metric.Builder()
```

Comment:
Perhaps create a base builder from `metric` to ensure any fields that you want copied over are even if the definition of `Metric` changes.

https://github.com/ArpNetworking/commons/blob/master/src/main/java/com/arpnetworking/commons/builder/ThreadLocalBuilder.java#L151

Reply:
Added a 'fromMetric' static method on the builder.